vowpalwabbit/namespaced_feature_store.cc

```diff
@@ -7,344 +7,89 @@
-{
-  return std::make_pair(namespace_index_begin(ns_index), namespace_index_end(ns_index));
-}
-
-features& namespaced_features::get_or_create_feature_group(uint64_t hash, namespace_index ns_index)
-{
-  auto* existing_group = get_feature_group(hash);
+  auto* existing_group = get_or_null(ns_index, hash);
   if (existing_group == nullptr)
   {
-    _feature_groups.emplace_back();
-    _namespace_indices.push_back(ns_index);
-    _namespace_hashes.push_back(hash);
-    auto new_index = _feature_groups.size() - 1;
-    _hash_to_index_mapping[hash] = new_index;
-    _legacy_indices_to_index_mapping[ns_index].push_back(new_index);
-    // If size is 1, that means this is the first time the ns_index is added and we should add it to the set.
-    if (_legacy_indices_to_index_mapping[ns_index].size() == 1) { _contained_indices.insert(ns_index); }
-    existing_group = &_feature_groups.back();
-  }
-
-  return *existing_group;
-}
-
-// This operation is only allowed in code that allows exceptions.
-// get_feature_group should be used instead for noexcept code
-#ifndef VW_NOEXCEPT
-const features& namespaced_features::operator[](uint64_t hash) const
-{
-  auto* existing_group = get_feature_group(hash);
-  if (existing_group == nullptr) { THROW("No group found for hash: " << hash); }
-  return *existing_group;
-}
-#endif
-
-#ifndef VW_NOEXCEPT
-features& namespaced_features::operator[](uint64_t hash)
-{
-  auto* existing_group = get_feature_group(hash);
-  if (existing_group == nullptr) { THROW("No group found for hash: " << hash); }
-  return *existing_group;
-}
-#endif
-
-void namespaced_features::remove_feature_group(uint64_t hash)
-{
-  if (_hash_to_index_mapping.count(hash) == 0) { return; }
-  auto existing_index = _hash_to_index_mapping[hash];
-
-  // Remove item from each vector at this index.
-  _feature_groups.erase(_feature_groups.begin() + existing_index);
-  _namespace_indices.erase(_namespace_indices.begin() + existing_index);
-  _namespace_hashes.erase(_namespace_hashes.begin() + existing_index);
-  _hash_to_index_mapping.erase(hash);
-
-  for (auto& kv : _legacy_indices_to_index_mapping)
-  {
-    auto& index_vec = kv.second;
-    // Remove this index from ns_index mappings if it exists
-    auto it = std::find(index_vec.begin(), index_vec.end(), existing_index);
-    if (it != index_vec.end()) { index_vec.erase(it); }
-
-    // Shift down any index that came after this one.
-    for (auto& idx : index_vec)
-    {
-      if (idx > existing_index) { idx -= 1; }
-    }
-  }
-
-  // If any groups are left empty, remove them.
-  for (auto it = _legacy_indices_to_index_mapping.begin(); it != _legacy_indices_to_index_mapping.end();)
-  {
-    if (it->second.empty())
+    if (!_saved_feature_group_nodes.empty())
     {
-      // There are no more feature groups which correspond to this index.
-      _contained_indices.erase(it->first);
-      it = _legacy_indices_to_index_mapping.erase(it);
+      _feature_groups[ns_index].splice(
```

Comment:
only the first element is being moved?

Reply:
Yep it is taking a single node and moving it over to be used